quick_blue_windows/windows/quick_blue_windows_plugin.cpp

```diff
@@ -461,6 +469,33 @@ winrt::fire_and_forget QuickBlueWindowsPlugin::WriteValueAsync(BluetoothDeviceAg
   OutputDebugString((L"WriteValueAsync " + winrt::to_hstring(characteristic) + L", " + winrt::to_hstring(to_hexstring(value)) + L", " + winrt::to_hstring((int32_t)writeValueStatus) + L"\n").c_str());
 }
 
+winrt::fire_and_forget QuickBlueWindowsPlugin::discoverServices(BluetoothDeviceAgent &bluetoothDeviceAgent, std::string deviceId) {
+    auto serviceResult = co_await bluetoothDeviceAgent.device.GetGattServicesAsync();
+    for (auto s : serviceResult.Services())
+    {
+      auto service = to_uuidstr(s.Uuid());
+      std::cout << service << std::endl;
+      auto characteristicResult = co_await s.GetCharacteristicsAsync();
```

Comment:
I've noticed these methods are prone to failure. In the winrt bindings HRESULTs other than success will throw an exception (you can look at base.h for an example).

Specifically I see
```        if (result == impl::error_illegal_state_change)
        {
            throw hresult_illegal_state_change(take_ownership_from_abi);
        }
```

Frequently for this method. I think it is because there is a change in the state of the device between when it was checked and when it is used. (An unavoidable situation.) 

So I think this, and other such calls, require exception handling.
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/error-handling#catching-exceptions

I am not a maintainer for this package, but I observed this in testing an implementation similar to this I was working on yesterday.

Reply:
Any workaround for this logic?

Reply:
As far as I am aware the only way, while also using the WinRT binding, would be to try/catch around these method calls. There may be some way to allow you to handle the hresult directly instead, but I am not familiar enough with WinRT to know.

I suspect exception handling to be the easiest approach here though, as this same problem even applies to the C# version of these APIs.

The most succinct approach would be to wrap each of the windows APIs in a thin wrapper. 

```
  IAsyncOperation<GattDeviceServiceResult> GetGattServicesAsync(BluetoothLEDevice device) {
        try {
            auto serviceResult = co_await device.GetGattServicesAsync();

            if (serviceResult.Status() != GattCommunicationStatus::Success)
                co_return nullptr;

			return serviceResult

		// The catch could be winrt::hresult_error const& ex, but I am not sure how helpful being specific would be here.
        } catch (...) {
            co_return nullptr;
        }
    }
```

This is just pseudocode, but then you could error handle the other items like normal.

There are a couple more cases that are problematic. Though they aren't directly in this PR.
```
co_await bluetoothDeviceAgent.GetCharacteristicAsync(service, characteristic);
```
Like the device from address, this can return null.

```
co_await gattCharacteristic.WriteClientCharacteristicConfigurationDescriptorAsync(descriptorValue);
```

This method seems to throw an exception if the descriptor isn't supported. Like requesting notify for a characteristic that only supports indicate.

Reply:
I don't think you can avoid the possibility of error. The connections can always change while you are doing operations. Unfortunately windows handles this pretty poorly in the interface.

Reply:
Yep. A lot of issues caused by WinRT exceptions. PRs welcome

Reply:
@kinyoklion can you push a commit to this one for this fix

Reply:
@rohitsangwan01 I made a PR against your branch in your fork with the changes. https://github.com/rohitsangwan01/quick_blue/pull/1

@Sunbreak If that approach is good for you, then I can make another PR with remaining hardening. It would be more of the same, just taking the WinRT method, writing a free function with exception handling, and then replacing calls. Also some null checks for things that can be null.

I do have one more question about this PR specifically. The way I was using this method was to determine the capabilities of a device before I connected to it. Being as I only want list/connect to devices which support specific functionality. This PR requires connecting to the devices before discovering services. Is this the intention for the API? It looks like the swift implementation allows discovering services for discovered devices (versus connected).

I do have a suggestion, which I could also implement in another PR, which would be to add the advertised services (those in the advertisement data) to the scan results. The advertised services are generally enough to know if a device is of interest, and as far as I am aware all platforms have a way to get it from their API.